python/ray/_private/runtime_env/working_dir.py

```diff
@@ -317,128 +299,164 @@ def package_exists(pkg_uri: str) -> bool:
+                _logger.info(
+                    f"{pkg_uri} doesn't exist. Create new package with"
+                    f" {working_dir} and {py_modules}")
+                if not pkg_file.exists():
+                    create_project_package(working_dir, py_modules, excludes,
+                                           file_path)
+                # Push the data to remote storage
+                pkg_size = push_package(pkg_uri, pkg_file)
+                _logger.info(
+                    f"{pkg_uri} has been pushed with {pkg_size} bytes")
+
+    def ensure_runtime_env_setup(self, pkg_uris: List[str]) -> Optional[str]:
+        """Make sure all required packages are downloaded it local.
+
+        Necessary packages required to run the job will be downloaded
+        into local file system if it doesn't exist.
+
+        Args:
+            pkg_uri list(str): Package of the working dir for the runtime env.
+
+        Return:
+            Working directory is returned if the pkg_uris is not empty,
+            otherwise, None is returned.
+        """
+        pkg_dir = None
+        for pkg_uri in pkg_uris:
+            # For each node, the package will only be downloaded one time
+            # Locking to avoid multiple process download concurrently
+            pkg_file = Path(self._get_local_path(pkg_uri))
+            with FileLock(str(pkg_file) + ".lock"):
+                pkg_dir = self.fetch_package(pkg_uri)
+            sys.path.insert(0, str(pkg_dir))
+
+        # Right now, multiple pkg_uris are not supported correctly.
+        # We return the last one as working directory
+        return str(pkg_dir) if pkg_dir else None
+
+    def delete_uri(self, uri: str) -> bool:
+        """Deletes a specific URI from the local filesystem.
+
+        Args:
+            uri (str): URI to delete.
+
+        Returns:
+            True if the URI was successfully deleted, else False.
+        """
+        deleted = False
+        path = Path(self._get_local_path(uri))
+        with FileLock(str(path) + ".lock"):
+            path = path.with_suffix("")
+            if path.exists():
+                if path.is_dir() and not path.is_symlink():
+                    shutil.rmtree(str(path))
+                else:
+                    path.unlink()
+                deleted = True
+
+        if not deleted:
+            _logger.warning(f"Tried to delete nonexistent path: {path}")
+
+        return deleted
+
+    def setup(self,
+              runtime_env: dict,
+              context: RuntimeEnvContext,
+              logger: Optional[logging.Logger] = None):
+        if not runtime_env.get("uris"):
+            return
+
+        # Overwrite the module-wide logger temporarily.
+        # TODO(edoakes): we should be able to remove this by refactoring the
+        # working_dir setup code into a class instead of using global vars.
+        global _logger
+        if logger:
+            prev_logger = _logger
+            _logger = logger
```

Comment:
It looks like with the class introduced in this PR, we can remove this now, but it should be fine to do it in a future PR.

Reply:
Yeah we should definitely be able to remove this now, will do it when I clean up the class structure!